### ISSUE TITLE
High CPU usage when using 3131

### ISSUE THREAD

**Issue by javorszky** [equinusocio/material-theme#1128]
### Summary

Since updating to 3131, the energy impact / cpu used by Sublime has been constantly very high. Reverting to a freshly installed state helped. Once I reinstalled Material Theme, the CPU troubles began again. Once I reset the theme to be default, the CPU troubles went away.

### Expected behavior

Using Material theme shouldn't be taxing on the CPU.

### Actual behavior

Using Material theme is taxing on the CPU.

### Steps to reproduce

1. Have a freshly installed copy of Sublime text 3131
2. Install and activate Material theme
3. Observe the high CPU load.

### Environment

* Operating system and version:
  * [x] Mac OS 10.12.4
* Sublime Text:
  * Build 3131
* Material Theme release:
  * 4.0.6

Info copied from Material Theme: Info

* Sublime Text:   3131
* Platform:   osx
* Arch:   x64
* Theme:   4.0.6
* Install via PC:   True
* Dependencies:
    * mdpopups:   1.13.1
    * markdown:   2.6.6
    * pygments:   2.1a0
    * jinja2:   2.8

**Comment by equinusocio** [comment]
I think there is another problem with your dependencies/packages. Please tell me more infos like packages installed

**Comment by javorszky** [comment]
Complete list of packages installed from `Packages/User/Package Control.sublime-settings`

```
{
	"bootstrapped": true,
	"in_process_packages":
	[
	],
	"installed_packages":
	[
		"DocBlockr",
		"EditorConfig",
		"Emmet",
		"Git",
		"GitGutter",
		"Material Theme",
		"Package Control",
		"Sass",
		"SideBarEnhancements"
	]
}
```

I have 4 folders open and 2 individual files. With Material theme the CPU usage jumps to 70%, then levels at 20%, and then occasionally return to 5% where it is with default theme.

**Comment by equinusocio** [comment]
Have you tried a fresh ST installation with only Material Theme?

**Comment by javorszky** [comment]
Hm, at the very least I'd need package control on. I'll try that, hold on a second.

**Comment by javorszky** [comment]
Yes, this is still happening.

* I deleted the `~/Library/Application Support/Sublime Text 3` folder
* then opened it, added my licence key
* installed package control
* installed material theme
* installed the file icon pack that popped up
* restarted sublime
* activated material theme
* opened a folder (wordpress theme) 

Here's a gif of my activity monitor and the settings file:

![screen recording 2017-05-03 at 11 52 am](https://cloud.githubusercontent.com/assets/1196299/25657702/199752f6-2ff7-11e7-926a-980823a3a689.gif)

**Comment by equinusocio** [comment]
Mmm i don't know how help you because i can't reproduce this issue on my mbp.

**Comment by javorszky** [comment]
Hm... I'll do a complete removal / reinstall of ST3... maybe something's corrupted in the app's files.

**Comment by javorszky** [comment]
Are you using the dev build (3131) or the release build (3126)?

**Comment by equinusocio** [comment]
Dev build 3131

**Comment by ihodev** [comment]
@javorszky Try to remove `Emmet`.

**Comment by equinusocio** [comment]
@ihodev Why emmet?

**Comment by javorszky** [comment]
@ihodev the second time I tried it after reinstalling sublime I didn't install any packages outside of package control, material theme, and the icon font pack. There was no emmet to remove.

**Comment by ihodev** [comment]
@equinusocio In some cases it can cause such issues.

**Comment by StephanDollberg** [comment]
Seeing the same issue on debian stretch / linux 4.9 / gnome 3.22 using sublime text 3131

https://forum.sublimetext.com/t/sublime-text-3131-hugging-5-cpu-while-idle/28141

Tried with clean install as well to no success.

**Comment by StephanDollberg** [comment]
I have posted this on the subl forum as well.

If you strace the process you see the following pattern

```
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}, {fd=5, events=POLLIN}, {fd=7, events=POLLIN}, {fd=12, events=POLLIN}, {fd=13, events=POLLIN}, {fd=14, events=POLLIN}, {fd=15, events=POLLIN}, {fd=18, events=POLLIN}], 9, 0) = 0 (Timeout)
recvmsg(3, {msg_namelen=0}, 0)          = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}, {fd=5, events=POLLIN}, {fd=7, events=POLLIN}, {fd=12, events=POLLIN}, {fd=13, events=POLLIN}, {fd=14, events=POLLIN}, {fd=15, events=POLLIN}, {fd=18, events=POLLIN}], 9, 0) = 0 (Timeout)
recvmsg(3, {msg_namelen=0}, 0)          = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}, {fd=5, events=POLLIN}, {fd=7, events=POLLIN}, {fd=12, events=POLLIN}, {fd=13, events=POLLIN}, {fd=14, events=POLLIN}, {fd=15, events=POLLIN}, {fd=18, events=POLLIN}], 9, 15) = 0 (Timeout)
write(8, "\0", 1)                       = 1
recvmsg(3, {msg_namelen=0}, 0)          = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}, {fd=5, events=POLLIN}, {fd=7, events=POLLIN}, {fd=12, events=POLLIN}, {fd=13, events=POLLIN}, {fd=14, events=POLLIN}, {fd=15, events=POLLIN}, {fd=18, events=POLLIN}], 9, 0) = 1 ([{fd=7, revents=POLLIN}])
read(7, "\0", 1)                        = 1
recvmsg(3, {msg_namelen=0}, 0)          = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}, {fd=5, events=POLLIN}, {fd=7, events=POLLIN}, {fd=12, events=POLLIN}, {fd=13, events=POLLIN}, {fd=14, events=POLLIN}, {fd=15, events=POLLIN}, {fd=18, events=POLLIN}], 9, 0) = 0 (Timeout)
recvmsg(3, {msg_namelen=0}, 0)          = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}, {fd=5, events=POLLIN}, {fd=7, events=POLLIN}, {fd=12, events=POLLIN}, {fd=13, events=POLLIN}, {fd=14, events=POLLIN}, {fd=15, events=POLLIN}, {fd=18, events=POLLIN}], 9, 0) = 0 (Timeout)
recvmsg(3, {msg_namelen=0}, 0)          = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}, {fd=5, events=POLLIN}, {fd=7, events=POLLIN}, {fd=12, events=POLLIN}, {fd=13, events=POLLIN}, {fd=14, events=POLLIN}, {fd=15, events=POLLIN}, {fd=18, events=POLLIN}], 9, 0) = 0 (Timeout)
recvmsg(3, {msg_namelen=0}, 0)          = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}, {fd=5, events=POLLIN}, {fd=7, events=POLLIN}, {fd=12, events=POLLIN}, {fd=13, events=POLLIN}, {fd=14, events=POLLIN}, {fd=15, events=POLLIN}, {fd=18, events=POLLIN}], 9, 15^Cstrace: Process 5663 detached
```
even if the window is not active with material theme.

Without material theme that pattern is only visible if subl is focused but strace is pretty silent if subl is not focused.

**Comment by pzgz** [comment]
Having exactly the same issue, on macOS Sierra, the `WindowServer` process always bumped up when I start sublime and activate the theme. It will burn up the fans as well. Have to disable the theme for now. Looking forward for a fix.

**Comment by sn0wyfall** [comment]
For me it seems to turn on the fans as well! I'm not 100% sure I can attribute it to this theme, though.

**Comment by mwean** [comment]
I'm having the same issue. 20-40% CPU at idle, and it goes away when I switch themes.

**Comment by 449134904** [comment]
在phpstorm 白色是有问题的

**Comment by 449134904** [comment]
在phpstorm 白色是有问题的，每次开启都要重新设置一次

**Comment by 449134904** [comment]
In phpstorm Material-theme-Lighter is a bug, each time you have to re-set once

**Comment by equinusocio** [comment]
Report this issue to the phpstorm material theme repository please.

Il mer 17 mag 2017, 04:54 449134904 <notifications@github.com> ha scritto:

**Comment by sampotts** [comment]
As odd as this sounds, I had the same issue - CPU usage for Sublime at 40%. Tried a few things including changing indexing settings etc which made no difference and then found this thread. I thought I'd try flicking to the default theme (Under `Settings > Theme...`) which dropped CPU down to 1 or 2% , then I went back to the material theme and low and behold only 1 or 2% usage again. It's been like that for an hour or two now.

**Comment by StephanDollberg** [comment]
Still seeing the same problem subl 3132

**Comment by processwired** [comment]
I also suffer from this cpu hunger and did a lot of testing to get an idea what may be the culprit.

First, I deleted every plugin and every setting from Sublime, with exception of
```
Installed Packages/Material Theme.sublime-package
Packages/User/Color Highlighter/themes/Material-Theme-Darker.tmTheme
Packages/User/Color Highlighter/themes/Material-Theme.tmTheme
```
Then I downloaded and tested all Sublime versions starting with 3126 to most current 3132. My results are: 
with version 3126 all is well, no high cpu intensity. But every more up to date version (3127 and newer) cpu usage is between 15 and 25 percent when Sublime is idle.

Only setting in this test cases was:
```
{
	"theme": "Material-Theme.sublime-theme"
}
```
**If I remove this setting, Sublime works without any noticeable cpu usage in every version**. So, my conclusion is that it must be related to the theme.

I hope this helps so far.

**Comment by sampotts** [comment]
For me on OSX and Windows, swapping the theme to default under the "Theme..." option and then back to Material theme seems to resolve the CPU usage for me. Very weird.

**Comment by equinusocio** [comment]
Can you enable the theme by using che comman `UI: Select Skin` and check if you have the same CPU issue?

**Comment by equinusocio** [comment]
Give me updates.

**Comment by sampotts** [comment]
Sorry about the delay. Using the command palette and using `UI: Select Theme` (I couldn't see `UI: Select Skin`) seemed to give the same result. I have to switch back to default and then to material theme and it's ok. I have to do it every time I restart. Before doing this theme swap Sublime (with material theme) uses ~30% CPU and after the swap ~10%. If Sublime doesn't have focus then CPU usage is <1 % usually.

**Comment by equinusocio** [comment]
@sampotts Can you try to install Skins package and then follows this steps:
- Set default theme
- Restart ST
- Set one of material theme with the `UI: Select Skins` command (provided with Skins package)
- Check the CPU
- Restart
- Check CPU again.

**Comment by sampotts** [comment]
First CPU check:
![image](https://user-images.githubusercontent.com/719092/27852809-e7bb008c-61a3-11e7-9259-44bb11770def.png)

Will restart and post again

**Comment by sampotts** [comment]
After restart back to 30% usage

**Comment by equinusocio** [comment]
I can't figure out where is the issue since i can't reproduce this, even after a fresh ST 3140 installation.

**Comment by sampotts** [comment]
Could well be down to another plugin then I guess? 

I have:

- Alignment
- CSSComb
- EditorConfig
- Emmet
- Emmet CSS Snippets
- HTML-JS-CSS Prettify
- JavaScriptNext - ES6 Syntax
- jQuery
- LESS
- Liquid
- Material Theme (of course)
- PackageControl
- PyV8
- Razor
- SASS
- SideBarEnhancements
- SublimeLinter
- SublimeLinter-jshint
- SublimeLinter-json

**Comment by equinusocio** [comment]
You should try a fresh installation and install your packages one by one testing the CPU after the theme selection.

**Comment by StephanDollberg** [comment]
This problem is definitely caused by material theme. As many have mentioned in this thread, I have now also long uninstalled the plugin and everything is fine since then.

It doesn't have to be bug in material theme but it certainly makes the bug surface.

**Comment by sn0wyfall** [comment]
Well, okay. Someone should try a fresh installation and then enable their own packages one by one or something like that. The way we're going, we're not helping @equinusocio at all, and he has no idea what's causing this. You all probably know how hard it is to fix a bug when you can't reproduce it – when you have no idea when that bug even arises or can be called. I'd do it myself but I don't seem to be facing this issue right now.

**Comment by equinusocio** [comment]
> I'd do it myself but I don't seem to be facing this issue right now.

This is also strange because if material theme generate an issue it should generate the same issue on every installation with the same os platform.

**Comment by sn0wyfall** [comment]
@equinusocio I have the material theme and a few Sublime Text windows open in the background. Sublime Text is using 0.00-0.02% of my CPU, hardly has any energy impact, and is "sleeping" on macOS. I'll definitely try to find out what causes it if it happens (again), and I'd ask someone else who's facing it to find please out what package (incompatibilities?) are causing it. Otherwise we're just complaining, not helping :) I hope that's okay

**Comment by equinusocio** [comment]
Make sure also to use ST 3140

**Comment by sampotts** [comment]
Mine uses < 1% when it's not in focus, as I said above. That's not the issue though, or at least not for me. It's when Sublime *has* focus it uses ~30% CPU. 

I wasn't complaining 👍 I'm trying to help by listing my OS and possible problematic plugins.

**Comment by sn0wyfall** [comment]
@sampotts Oh no, I totally, understand, I'm just telling everyone in general :). For me, when I'm not doing anything, Sublime Text takes around 3.5% of the CPU without the theme, and around 2.5% with the theme (of course there's some margin of error), typing takes 10-15% without and 10-18% with, scrolling quickly takes 40% without and 30-55% with (these are very non-scientific as you can see!), so in general it doesn't seem to be a problem to inherent to the theme as the author confirmed.

**Comment by equinusocio** [comment]
Since of lack of feedbacks i will close this for now.

**Comment by sampotts** [comment]
I've since switched to Visual Studio Code (which I much prefer and uses 0-1% CPU at idle). I'm also using the Material theme with that 👍

**Comment by StephanDollberg** [comment]
@equinusocio not sure which feedback you want more. It's up to you to fix the bug now. Closing this ticket won't make it go away.

**Comment by mallowigi** [comment]
Removing the gutters might be a fix

**Comment by equinusocio** [comment]
@StephanDollberg Why you don't clone and try to fix the issue? Please remove this theme.

**Comment by ximus** [comment]
happy to help debug this.

here is an osx process sample during high cpu activity: https://gist.github.com/ximus/1fb5ef865a8cd263d6aaefcd979ed27d

i'm not familiar with Sublime internals, but looks like graphics/text-related threads are busy during that sample window

let me know if you want me try something

**Comment by equinusocio** [comment]
@ximus thank you. I have no idea about how to fix it because i don’t have this issue..

Can you try to extract the package in your packages folder and remove all python files?

**Comment by ximus** [comment]
hoping i understood you right, this is what i did:

within `Installed Packages`, I extracted `Material Theme.sublime-package` to `Packages/Material Theme`. Inside the extract i removed:
```
> find . -iname "*.py"
./Icons.py
./MT.py
./test/source.py
./utils/__init__.py
./utils/activation.py
./utils/changelog.py
./utils/config.py
./utils/extras.py
./utils/info.py
./utils/links.py
```

then i started and restarted sublime, but CPU was still same 10-12% usage for a single window with no open files (vs 0-1% usage when theme is disabled).

I also tried removing `Installed Packages/Material Theme.sublime-package` while keeping `Packages/Material Theme`. Theme still worked but CPU usage was still high.

**Comment by ximus** [comment]
i'm still running with the extracted `Packages/Material Theme` without the python files folder. Now i'm seeing CPU at ~4%. Not sure what to think at this point. I'm going to leave it like this and work with it for a day or so, I'll let you know if i see any improvements from this.

**Comment by equinusocio** [comment]
@ximus can you try to remove Material Theme and Install DA theme and then check if the CPU still high?

**Comment by equinusocio** [comment]
@ximus removing the python files you are not able to use the material theme command palette.

**Comment by mallowigi** [comment]
I've been running with the packages in my disk and didn't get a CPU increase, so I guess it's not directly related...

**Comment by ihodev** [comment]
All that I can say, this issue is not related to python stuff (it should work as expected). Material Theme and Boxy Theme have too many styling overrides (new ST engine doesn't like that), introduced by theme settings. The only possible way to fix this is rewrite theme from scratch according to [ST docs](https://www.sublimetext.com/docs/3/themes.html).

Also there is [the issue](https://github.com/ihodev/a-file-icon/issues/133) with `A File Icon`, which can make these themes a little bit slower. Currently, I'm working on [the new version](https://github.com/ihodev/a-file-icon/issues/139) and the fix will be available in it.

**Comment by equinusocio** [comment]
And.. the ST theme engine is overcomplicated and hard to mantain considering the VSC theme engine where you can do less but better.

**Comment by betorobson** [comment]
I would like to share with you what I catch form activity monitor from Mac MacBook Pro 2017 (no touch bar). I'm using a external monitor 4k 2160p scaled resolution to 2560x1440.

I did a comparison between Sublime and VSCode, I really love Sublime, I want keep using it but as long as it has a high energy impact I'd rather starting using VSCode until this issue get fixed.

Sorry for my accent! :-)

Hope those assets can help you guys.

I have 1 screen recorder and 2 screenshot of both editors opened.

https://www.youtube.com/watch?v=J8InxhjS6IY&feature=youtu.be

<img width="2672" alt="screen shot 2017-10-15 at 8 29 14 pm" src="https://user-images.githubusercontent.com/13212208/31589983-51e989ee-b1e8-11e7-8289-4ebfabd27f83.png">
<img width="2672" alt="screen shot 2017-10-15 at 8 29 23 pm" src="https://user-images.githubusercontent.com/13212208/31589984-520918fe-b1e8-11e7-9584-cf5d3c20955c.png">

**Comment by equinusocio** [comment]
@betorobson consider that this theme is not under active development. I’m now focused on Material Theme for visual studio code and if you want to keep using it you can find it on the vsc marketplace. Be aware from clones. 

https://github.com/equinusocio/vsc-material-theme

**Comment by kawasako** [comment]
If we give up the style of titlebar this problem seems to be solved. Please give us your opinion about this.

**Comment by chanpory** [comment]
I also experiencing extreme high CPU usage, whenever I have Material Theme enabled on Build 3143. This started happening shortly after upgrading to Sublime Text 3.0. 

Even after a fresh install of Sublime 3 and reinstalling the theme causes massive slowdown on my MacBook Pro Touchbar.

The only error that I see in console is:

```
Errors parsing theme:
icon_file_type is missing layer0.opacity, setting to 1.0 for backwards compatibility
reloading settings Packages/User/Preferences.sublime-settings
```

**Comment by chanpory** [comment]
Putting `gpu_window_buffer": false` into settings seems to be helping a bit.

**Comment by raulp** [comment]
This issue still persists.
What works for me currently is ... after sublime boots, to open preferences, disable the theme and enable it back. Then my cpu usage is indeed < 1%, 0 when idle.

**Comment by sixthgear** [comment]
Seeing this issue as well on 10.11.6.

**Comment by betorobson** [comment]
I'm following the @ihodev suggestion which tell me the Material Theme is not supported on Sublime anymore so, I move on to DA UI https://github.com/ihodev/sublime-da-ui.
It is working a little bit better than Material Theme.

**Comment by ihodev** [comment]
I did not say that it's not supported anymore, but it requires significant rewrite for better performance :(
This one of the reasons why I've decided to deprecate Boxy Theme.

**Comment by betorobson** [comment]
OMG!!! Really Really Sorry, I made a mistake, it was a confusion. My Bad!!!
You are completely right, in another thread we have talked about the Boxy Theme.

I'm so shame of me, sorry :-(

Guys, what happens why that I replace Material Theme due to high energy impact by Boxy Theme and than @ihodev  tell me would be a batter options move on to DA UI.

On more time again.... Sorry about my confusion.
